renderdoc/driver/d3d8/d3d8_resources.h

```diff
@@ -0,0 +1,250 @@
+#include "d3d8_manager.h"
+
+enum ResourceType
+{
+  Resource_Unknown = 0,
+  Resource_VertexBuffer,
+  Resource_IndexBuffer,
+};
+
+ResourceType IdentifyTypeByPtr(IDirect3DResource8 *ptr);
+ResourceId GetIDForResource(IDirect3DResource8 *ptr);
+
+class TrackedResource
+{
+public:
+  TrackedResource() { m_ID = ResourceIDGen::GetNewUniqueID(); }
+  ResourceId GetResourceID() { return m_ID; }
+private:
+  TrackedResource(const TrackedResource &);
+  TrackedResource &operator=(const TrackedResource &);
+
+  ResourceId m_ID;
+};
+
+template <typename NestedType>
+class WrappedResource8 : public RefCounter8, public NestedType, public TrackedResource
+{
+protected:
+  WrappedD3DDevice8 *m_pDevice;
+  NestedType *m_pReal;
+  IID m_NestedTypeIID;
+  unsigned int m_PipelineRefs;
+
+  WrappedResource8(NestedType *real, WrappedD3DDevice8 *device, IID nestedTypeIID)
+      : RefCounter8(real),
+        m_pDevice(device),
+        m_NestedTypeIID(nestedTypeIID),
+        m_pReal(real),
+        m_PipelineRefs(0)
+  {
+    m_pDevice->SoftRef();
+
+    bool ret = m_pDevice->GetResourceManager()->AddWrapper(this, real);
+    if(!ret)
+      RDCERR("Error adding wrapper for type %s", ToStr::Get(m_NestedTypeIID).c_str());
+
+    m_pDevice->GetResourceManager()->AddCurrentResource(GetResourceID(), this);
+  }
+
+  virtual void Shutdown()
+  {
+    m_pDevice->GetResourceManager()->RemoveWrapper(m_pReal);
+    m_pDevice->GetResourceManager()->ReleaseCurrentResource(GetResourceID());
+    m_pDevice->ReleaseResource((NestedType *)this);
+    SAFE_RELEASE(m_pReal);
+    m_pDevice = NULL;
+  }
+
+  virtual ~WrappedResource8()
+  {
+    // should have already called shutdown (needs to be called from child class to ensure
+    // vtables are still in place when we call ReleaseResource)
+    RDCASSERT(m_pDevice == NULL && m_pReal == NULL);
+  }
+
+public:
+  typedef NestedType InnerType;
+
+  NestedType *GetReal() { return m_pReal; }
+  ULONG STDMETHODCALLTYPE AddRef() { return RefCounter8::SoftRef(m_pDevice) - m_PipelineRefs; }
+  ULONG STDMETHODCALLTYPE Release()
+  {
+    unsigned int piperefs = m_PipelineRefs;
+    return RefCounter8::SoftRelease(m_pDevice) - piperefs;
+  }
+
+  void PipelineAddRef() { InterlockedIncrement(&m_PipelineRefs); }
+  void PipelineRelease() { InterlockedDecrement(&m_PipelineRefs); }
+  HRESULT STDMETHODCALLTYPE QueryInterface(REFIID riid, void **ppvObject)
+  {
+    if(riid == __uuidof(IUnknown))
+    {
+      *ppvObject = (IUnknown *)(NestedType *)this;
+      AddRef();
+      return S_OK;
+    }
+    if(riid == m_NestedTypeIID)
+    {
+      *ppvObject = (NestedType *)this;
+      AddRef();
+      return S_OK;
+    }
+
+    // DEFINE_GUID(IID_IDirect3DResource8, 0x1b36bb7b, 0x9b7, 0x410a, 0xb4, 0x45, 0x7d, 0x14, 0x30,
+    // 0xd7, 0xb3, 0x3f);
+    static const GUID IDirect3DResource8_uuid = {0x1b36bb7b, 0x9b7, 0x410a, 0xb4, 0x45, 0x7d,
+                                                 0x14,       0x30,  0xd7,   0xb3, 0x3f};
```

Comment:
It seems like you're jumping through some hoops to get the uuids here because `__uuidof` won't work out of the box on D3D8 interfaces.

Since I don't think they're releasing any changes to D3D8 any time soon (:wink:) would you be opposed to making local patches to d3d8.h to add the `DECLSPEC_UUID` annotations for `__uuidof` ?